### PR TITLE
[WIP] Support NHWC format in SpatialBatchNorm

### PR DESCRIPTION
**Pull request by yangw1234**
## What changes were proposed in this pull request?

support NHWC format in SpatialBatchNorm

## How was this patch tested?

add unit tests

### CODE REVIEW

**Comment by yangw1234** [comment]
pr validation pass

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala [hunk 1]
```diff
@@ -62,9 +439,10 @@ object SpatialBatchNormalization {
       initWeight: Tensor[T] = null,
       initBias: Tensor[T] = null,
       initGradWeight: Tensor[T] = null,
-      initGradBias: Tensor[T] = null)(implicit ev: TensorNumeric[T])
+      initGradBias: Tensor[T] = null,
+      format: DataFormat = DataFormat.NCHW)(implicit ev: TensorNumeric[T])
   : SpatialBatchNormalization[T] = {
     new SpatialBatchNormalization[T](nOutput, eps, momentum, affine,
-      initWeight, initBias, initGradWeight, initGradBias)
+      initWeight, initBias, initGradWeight, initGradBias, format)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala]: update the md document?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala [hunk 2]
```diff
@@ -51,6 +54,380 @@ class SpatialBatchNormalization[T: ClassTag](
...
+  private val saveVar = Tensor[T](nOutput)
+
+  private val ones = if (!affine && format.isInstanceOf[DataFormat.NHWC.type ]) {
+    Tensor[T](nOutput)
+  } else {
+    null
+  }
+
+  override protected def updateOutputDouble(
+                                             inputTensor: Tensor[Double],
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala]: indent

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala [hunk 3]
```diff
@@ -51,6 +54,380 @@ class SpatialBatchNormalization[T: ClassTag](
...
+                                             outputTensor: Tensor[Double]): Unit = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala]: indent

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala [hunk 4]
```diff
@@ -51,6 +54,380 @@ class SpatialBatchNormalization[T: ClassTag](
...
+    format match {
+      case DataFormat.NCHW =>
+        super.updateOutputDouble(inputTensor, outputTensor)
+      case DataFormat.NHWC =>
+        updateOutputDoubleNHWC(inputTensor, outputTensor)
+    }
+  }
+
+  override protected def updateOutputFloat(
+                                            inputTensor: Tensor[Float],
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala]: indent

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala [hunk 5]
```diff
@@ -51,6 +54,380 @@ class SpatialBatchNormalization[T: ClassTag](
...
+                                            outputTensor: Tensor[Float]): Unit = {
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala]: indent

File: spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalizationSpec.scala [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+package com.intel.analytics.bigdl.nn
+
+import com.intel.analytics.bigdl.nn.abstractnn.DataFormat
+import com.intel.analytics.bigdl.tensor.Tensor
+import org.scalatest.{FlatSpec, Matchers}
+
+@com.intel.analytics.bigdl.tags.Parallel
+class SpatialBatchNormalizationSpec extends FlatSpec with Matchers {
+
+  "SpatialBatchNormalization with NHWC and NCHW format" should "have the same result" in {
```
Comment (yiheng) [inline spark/dl/src/test/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalizationSpec.scala]: can you do some micro-benchmark for the two type of bsn?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala [hunk 7]
```diff
@@ -51,6 +54,380 @@ class SpatialBatchNormalization[T: ClassTag](
...
   }
+
+  override val batchFeatDim: Int = {
+    format match {
+      case DataFormat.NCHW => 2
+      case DataFormat.NHWC => 4
+    }
+  }
+
+  private val varBuffer = Tensor[T](nOutput)
```
Comment (yiheng) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/SpatialBatchNormalization.scala]: clear them in clearState?

**Comment by yiheng** [comment]
This has been fixd in PR#1936

**Pull request closed without merging**